### PR TITLE
[2.0] msal 2.0 supports throttling

### PR DESCRIPTION
**Pull request by jmckennon**
This PR implements the primary portion of client-side throttling in `msal-common`. There will be a few accompanying PRs associated with tests and `msal-browser`.

### CODE REVIEW

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (jmckennon) [inline lib/msal-common/src/network/ThrottlingManager.ts:38]: Ideally, we could make `preProcess` and `postProcess` static to allow other libraries to consume them without passing arround a `ThrottlingManager`, but I couldn't figure out a good way of accomplishing that as we also need to call the abstract functions here. Any thoughts?

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (pkanher617) [inline lib/msal-common/src/network/ThrottlingManager.ts:19]: Not sure I like calling this `ThrottlingManager`, would rather keep it as `NetworkManager`. IMO throttling is an implementation detail of our network class.

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (pkanher617) [reply lib/msal-common/src/network/ThrottlingManager.ts:38]: I don't think you need to make this static. You could make them protected.

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (tnorling) [reply lib/msal-common/src/network/ThrottlingManager.ts:19]: +1 I think `NetworkManager` should stay as is and call the throttling functions when needed

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (tnorling) [reply lib/msal-common/src/network/ThrottlingManager.ts:38]: I'm still unclear on why this needs to be abstract. Each library should provide their own `cacheManager` which you can use without needing each library to implement their own `ThrottlingManager`.

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingManager.ts:19]: @tnorling what do you consider "throttling functions"? the ones that are currently abstract?

File: lib/msal-common/src/network/RequestThumbprintValue.ts
Comment (tnorling) [inline lib/msal-common/src/network/RequestThumbprintValue.ts:25]: This function should be part of your Throttling class and just define a type in this file

File: lib/msal-common/src/network/RequestThumbprint.ts
Comment (tnorling) [inline lib/msal-common/src/network/RequestThumbprint.ts:21]: Same comment here

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (tnorling) [reply lib/msal-common/src/network/ThrottlingManager.ts:19]: Anything that is for the purpose of throttling, e.g. `preProcess`, `postProcess`, `getItem`, `setItem`,etc.

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (pkanher617) [reply lib/msal-common/src/network/ThrottlingManager.ts:38]: Since they are public or protected sub-classes can implement them and overwrite functionality

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingManager.ts:19]: So then the point of `NetworkManager` existing at all would be exclusively to hook into pre/post API call? Because at that point, the only method in the class is `sendPostRequest`. I kind of feel like it can be combined with what's in `INetworkModule` at that point, if I'm understanding correctly

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingManager.ts:38]: I think what Thomas is saying is that we don't necessarily need this to be abstract if we just directly use the provided cacheManager instead of forcing each library to implement it's own storage relative to throttling. cacheManager should be generic. Node just isn't set up to do that today, it would need to be incremental.

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (tnorling) [reply lib/msal-common/src/network/ThrottlingManager.ts:19]: Yeah, I suppose since the `NetworkManager` isn't implemented today you could just put the calls to `preProcess` and `postProcess` in the library implementations of the `INetworkModule`

File: lib/msal-common/src/client/AuthorizationCodeClient.ts
Comment (pkanher617) [inline lib/msal-common/src/client/AuthorizationCodeClient.ts:135]: should use the authority passed into the function instead of the request.authority here.

Comment (pkanher617) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:135]: `authority.canonicalAuthority`

File: lib/msal-common/src/client/BaseClient.ts
Comment (pkanher617) [inline lib/msal-common/src/client/BaseClient.ts:38]: change to network manager.

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingManager.ts]: Ended up moving these into a separate file, no longer have any abstract functions and we're directly using `cacheManager`.

File: lib/msal-common/src/network/ThrottlingManager.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingManager.ts]: I've kept `NetworkManager` separate for now as to not change the INetworkModule interface and to keep the implementation in common.

File: lib/msal-common/src/network/RequestThumbprint.ts
Comment (jmckennon) [reply lib/msal-common/src/network/RequestThumbprint.ts]: Done

File: lib/msal-common/src/client/AuthorizationCodeClient.ts
Comment (jmckennon) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts]: Done

File: lib/msal-common/src/client/BaseClient.ts
Comment (jmckennon) [reply lib/msal-common/src/client/BaseClient.ts]: Done

File: lib/msal-common/src/network/RequestThumbprintValue.ts
Comment (jmckennon) [reply lib/msal-common/src/network/RequestThumbprintValue.ts]: Done

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (tnorling) [inline lib/msal-common/src/network/ThrottlingUtils.ts:63]: Should this just be?
```suggestion
                JSON.stringify(thumbprintValue)
```

File: lib/msal-common/src/utils/StringUtils.ts
Comment (tnorling) [inline lib/msal-common/src/utils/StringUtils.ts:85]: Return null here?

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (tnorling) [inline lib/msal-common/src/network/ThrottlingUtils.ts:42]: To account for possibility of `jsonParseHelper` failing:
```suggestion
            if (parsedValue && parsedValue.throttleTime >= Date.now()) {
```

File: lib/msal-common/src/utils/Constants.ts
Comment (tnorling) [inline lib/msal-common/src/utils/Constants.ts:57]: Does this need to be an enum? Are there going to be more items under this?

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (tnorling) [inline lib/msal-common/src/network/ThrottlingUtils.ts:12]: These types should each be in their own separate files

Comment (jasonnutter) [reply lib/msal-common/src/network/ThrottlingUtils.ts:12]: Type only exports end up requiring typescript 3.9 in the client application, so we're trying to avoid that (at least for now).

Comment (tnorling) [reply lib/msal-common/src/network/ThrottlingUtils.ts:12]: Is that relevant for types that aren't considered publicly facing? We already have many Type only exports across both common and browser

Comment (jasonnutter) [reply lib/msal-common/src/network/ThrottlingUtils.ts:12]: Ah, that may not be the case for types that aren't part of the main export.

Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingUtils.ts:13]: When we say "public" here, do we mean being consumed by browser/node, or actually visible to customers? Because at least browser will need to use these to cover a few cases, I think

File: lib/msal-common/src/utils/StringUtils.ts
Comment (jmckennon) [reply lib/msal-common/src/utils/StringUtils.ts]: Done

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingUtils.ts]: Done

File: lib/msal-common/src/utils/Constants.ts
Comment (jmckennon) [reply lib/msal-common/src/utils/Constants.ts]: I ended up just putting this with the enum above it, HeaderNames. We can extract separate it for readability if needed

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingUtils.ts]: Done

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingUtils.ts]: I ended up putting these in separate files.

File: lib/msal-browser/src/cache/BrowserStorage.ts
Comment (jmckennon) [inline lib/msal-browser/src/cache/BrowserStorage.ts:26]: I don't really like this change, but I needed a way to obtain the `clientId` in the interaction handler class after a redirect and before calling into the `acquireToken` APIs. Any thoughts?

Comment (jmckennon) [reply lib/msal-browser/src/cache/BrowserStorage.ts:26]: This usage can be seen in the other changes to msal-browser in this PR

Comment (jasonnutter) [reply lib/msal-browser/src/cache/BrowserStorage.ts:26]: This technically makes it a public setter. I would keep it private then add a public getter function (e.g. getClientId)

**Comment by jasonnutter** [comment]
I think we might want to wait for after GA to merge and release this, thoughts @pkanher617? Should add to appropriate milestone.

File: lib/msal-browser/src/cache/BrowserStorage.ts
Comment (tnorling) [reply lib/msal-browser/src/cache/BrowserStorage.ts:26]: You should be able to get it from the cached request in the `RedirectHandler`

File: lib/msal-common/src/client/BaseClient.ts
Comment (tnorling) [inline lib/msal-common/src/client/BaseClient.ts:88]: Why is this added here and in `createDefaultTokenRequestHeaders`?

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (tnorling) [inline lib/msal-common/src/network/ThrottlingUtils.ts:32]: What cacheSchema does this fall under? Temporary? Throttling?

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (tnorling) [inline lib/msal-common/src/network/ThrottlingUtils.ts:73]: FYI I implemented this function slightly differently in Telemetry and included it in the NetworkManager class. Take a look at that PR and see if that might overlap with this here

File: lib/msal-common/src/utils/Constants.ts
Comment (tnorling) [inline lib/msal-common/src/utils/Constants.ts:41]: Should we add a separate Throttling object in this file for your constants?

File: lib/msal-common/src/client/BaseClient.ts
Comment (jmckennon) [reply lib/msal-common/src/client/BaseClient.ts:88]: This is a new header the server wants to receive for each request to basically tell the server that we support throttling. It's outlined in the original spec.

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingUtils.ts:32]: Thanks for this, had forgotten about the cache schema stuff, it's temporary. Changing

File: lib/msal-common/src/client/BaseClient.ts
Comment (tnorling) [reply lib/msal-common/src/client/BaseClient.ts:88]: Sure but the question is, why is it being added twice? Once as part of the default library headers and again as part of the default token request headers?

File: lib/msal-common/src/utils/Constants.ts
Comment (jmckennon) [reply lib/msal-common/src/utils/Constants.ts:41]: Done

File: lib/msal-browser/src/cache/BrowserStorage.ts
Comment (jmckennon) [reply lib/msal-browser/src/cache/BrowserStorage.ts]: I created a `getClientId()` for now. @tnorling It looks like the cached request is an `AuthorizationCodeRequest` which doesn't contain `clientId`, is there something else?

Comment (tnorling) [reply lib/msal-browser/src/cache/BrowserStorage.ts]: Ah you're right. Thought it was part of `BaseAuthRequest`. You could also pass it in from `PublicClientApplication` since it will be part of the config, but not sure what's cleaner

Comment (jmckennon) [reply lib/msal-browser/src/cache/BrowserStorage.ts]: Going to leave the `getClientId()` for now.

File: lib/msal-common/src/network/ThrottlingUtils.ts
Comment (jmckennon) [reply lib/msal-common/src/network/ThrottlingUtils.ts:74]: They're slightly different use cases, though both check against `status`. I'm going to leave this as-is for now.

File: lib/msal-browser/src/cache/BrowserStorage.ts
Comment (tnorling) [inline lib/msal-browser/src/cache/BrowserStorage.ts:427]: I still don't love this being part of the Storage class since it doesnt really have anything to do with storage. Would rather the clientId get passed into the RedirectHandler if thats the only place this is used

File: lib/msal-common/src/client/BaseClient.ts
Comment (tnorling) [reply lib/msal-common/src/client/BaseClient.ts:88]: Question still stands. This headers is added on Line 72 and on 88

Comment (jmckennon) [reply lib/msal-common/src/client/BaseClient.ts:88]: Sorry didn't see the reply here--I assumed this was two different places where headers get created, or does the token request headers just get added to the default library headers?

Comment (jmckennon) [reply lib/msal-common/src/client/BaseClient.ts:88]: On second though we probably only need them to be on the token requests anyway, as that's the only time we're looking to throttle a new thumbprint. Will change

Comment (jmckennon) [reply lib/msal-common/src/client/BaseClient.ts]: Changed, it now takes clientId as a parameter and uses that instead

File: lib/msal-browser/src/cache/BrowserStorage.ts
Comment (jmckennon) [reply lib/msal-browser/src/cache/BrowserStorage.ts]: Done

File: lib/msal-browser/src/interaction_handler/RedirectHandler.ts
Comment (tnorling) [inline lib/msal-browser/src/interaction_handler/RedirectHandler.ts:48]: Should this just be a required parameter?

Comment (jmckennon) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:48]: I don't think it can be since handleCodeResponse overrides the handleCodeResponse in InteractionHandler which does not specify clientId as a required parameter

Comment (jmckennon) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:48]: We could technically just rename the function here I think and be able to do that, though

Comment (tnorling) [reply lib/msal-browser/src/interaction_handler/RedirectHandler.ts:48]: Ahh that's right. No worries then

**Review by pkanher617 — APPROVED** [review]

**Pull request merged**
